validator/src/journal/chain_commit_state.rs

```diff
@@ -53,171 +43,78 @@ fn check_no_duplicates(ids: &[String]) -> Option<String> {
     None
 }
 
-impl<'b, 't, B: BatchIndex + 'b, T: TransactionIndex + 't> ChainCommitState<'b, 't, B, T> {
-    pub fn new<BS: BlockStore>(
-        branch_head_id: &str,
-        block_manager: &BlockManager,
-        batch_index: &'b B,
-        transaction_index: &'t T,
-        block_store: &BS,
-    ) -> Result<Self, ChainCommitStateError> {
-        let mut uncommitted_batch_ids = vec![];
-        let mut uncommitted_txn_ids = vec![];
-        let mut common_ancestor = None;
-        if branch_head_id != NULL_BLOCK_IDENTIFIER {
-            for block in block_manager.branch(branch_head_id).map_err(|err| {
-                ChainCommitStateError::Error(format!(
-                    "Error getting branch from block manager: {:?}",
-                    err
-                ))
-            })? {
-                let chain_head = block_store
-                    .get(&[&block.header_signature])
-                    .map_err(|err| {
-                        ChainCommitStateError::Error(format!(
-                            "Error getting block from blockstore: {:?}",
-                            err
-                        ))
-                    })?.next();
-
-                if chain_head.is_some() {
-                    common_ancestor = Some(block);
-                    break;
-                } else {
-                    for batch in &block.batches {
-                        uncommitted_batch_ids.push(batch.header_signature.clone());
-
-                        for transaction in &batch.transactions {
-                            uncommitted_txn_ids.push(transaction.header_signature.clone());
-                        }
-                    }
-                }
-            }
-        }
-        Ok(ChainCommitState {
-            batch_index,
-            transaction_index,
-            ancestor: common_ancestor,
-            uncommitted_batch_ids,
-            uncommitted_txn_ids,
-        })
-    }
-
-    fn block_in_chain(&self, block: &Block) -> bool {
-        if let Some(ref common) = self.ancestor {
-            return block.block_num <= common.block_num;
-        }
-        false
+pub fn validate_no_duplicate_batches(
+    block_manager: &BlockManager,
+    branch_head_id: &str,
+    batch_ids: Vec<String>,
+) -> Result<(), ChainCommitStateError> {
+    if let Some(batch_id) = check_no_duplicates(batch_ids.as_slice()) {
+        return Err(ChainCommitStateError::DuplicateBatch(batch_id));
     }
 
-    pub fn validate_no_duplicate_batches(
-        &self,
-        batch_ids: Vec<String>,
-    ) -> Result<(), ChainCommitStateError> {
-        if let Some(batch_id) = check_no_duplicates(batch_ids.as_slice()) {
+    for batch_id in batch_ids {
+        if block_manager
+            .contains_batch(branch_head_id, &batch_id)
+            .map_err(|err| ChainCommitStateError::Error(format!("{:?}", err)))?
```

Comment:
Is there not a `From<_>` implementation for this error type?  Since you're using `map_err` and a `?`and doing the same thing later, it might make sense, .